### ISSUE TITLE
Looking for someone with curating long-CoT dataset experience

### ISSUE THREAD

**Issue by Skimmable-Code-pls** [huggingface/open-r1#681]

**Comment by kiplangatkorir** [comment]
What specific aspects of long-CoT dataset curation you need help with?

**Comment by Skimmable-Code-pls** [comment]
@kiplangatkorir Thank you for your insights, they bring me a lot of assurance that I desperately need at the moment. 
>  Your reward function? Don't even think about embeddings yet. give it +1 for following the format, maybe +2 if it reuses appropriately, -1 if it makes a phrase that's basically identical to an existing one
Is there any chance I can tweak your proposed reward into a binary reward, such that I can utilise Dr.GRPO and BNPO which are known to work well only on binary reward from looking at `what isn't mentioned` in their experiment settings

> Also, be honest - do you actually have hundreds of papers with good ground truth phrases? Because if not, you're gonna need to make those first. And that's boring manual work, but it's the foundation of everything
I don't, my CEO told me that she doesn't trust the existing ground truth phrasing because her annotator was highly subjective, I think the close proxy to our work, and remaining consensus-ish, is Cohere's work on [Topic Clustering for Wikipedia page]. I believe one of their datasets [here](https://huggingface.co/Cohere/datasets) can be leveraged as the ground truth I desperately need. Though I need to investigate where their "Topics" were coming from.

> The growing phrase bank is clever, and the "reuse vs create new" decision is exactly the right framing
I'm afraid that this idea will bind my RL-tuning into 1 sample per batch so that the phrase bank faithfully increase/stay-the-same w.r.t Base Model generation. Am I correct to assume this?

**Comment by kiplangatkorir** [comment]
@Skimmable-Code-pls, you're absolutely right about the batch size thing - that's a good catch. If you're updating the phrase bank after each generation, you basically have to process papers one at a time. Otherwise you get weird race conditions where the model sees different phrase banks within the same batch.

But honestly? That might not be a dealbreaker. Sequential processing isn't ideal for throughput, but for research it's fine. You could also do some hacky workarounds like freezing the phrase bank for each batch, then updating it after the whole batch is done.

For the binary reward thing - yeah, totally doable. Just collapse your scoring into pass/fail:

```
Binary reward = 1 if:
- Follows format AND
- (Reuses appropriately OR creates genuinely new phrase)

Binary reward = 0 if:
- Doesn't follow format OR
- Creates redundant phrase OR  
- Misuses existing phrase
```

The tricky part is defining "reuses appropriately" and "genuinely new" in a way you can code up. But start simple - maybe just check if the new phrase has >80% word overlap with existing ones (redundant = bad).

The Cohere Wikipedia clustering thing is smart. I'd definitely dig into that - sounds way more reliable than subjective annotations. Plus if it's based on Wikipedia, your base model probably already knows those topics well.

Honestly, your CEO is right to be skeptical of the annotations. Subjective labeling without inter-annotator agreement is basically useless for training. Better to start with something more objective like the Cohere data, even if it's not perfectly matched to your domain.

One more thing - you might want to prototype this whole thing on a smaller model first. Way cheaper to debug your pipeline on a 7B model than burning money on something bigger while you're still figuring out the basics.

**Comment by Skimmable-Code-pls** [comment]
Thank you so so much <3!

**Comment by Skimmable-Code-pls** [comment]
One last question. I have a wild idea to leverage [Reinforcement Learned Teacher (RLT)](https://sakana.ai/rlt/) to eventually replace R1 as the Reference Model. From Dr.GRPO paper, I figured R1 tends to overthink, thus costing a lot of unnecessary GPU. What do you think?

**Comment by kiplangatkorir** [comment]
@Skimmable-Code-pls this is actually a really smart idea! I just read through the Sakana AI paper and it's way more clever than I initially gave it credit for.

The key thing is that RLT trains teachers to "learn to teach" rather than "learn to solve" - they get both the question AND the correct answer, then get rewarded based on how well a student model learns from their explanations. 

For your phrase generation task, this could be perfect:
- You'd give the RLT the paper data AND the target phrase
- It learns to explain WHY that phrase is the right choice
- Way more focused than R1's general reasoning chains

The efficiency gains are legit too - their 7B teacher outperformed DeepSeek R1 (671B params) on benchmarks: 26.3% vs 18.9%. That's not marginal, that's huge.

Your intuition about R1 overthinking is spot on. The paper mentions R1's outputs "often appear to rely on external tools like calculators, and even include out-of-place language patterns such as humorous comments" while RLT explanations are "more focused".

For GRPO, this could work really well because instead of R1 doing complex reasoning about "should I reuse this phrase or create new," you'd have a specialized teacher that's learned the specific patterns of good phrase selection.

Only caveat is you're adding another training step - you'd need to train the RLT first before using it in GRPO. But given the efficiency gains, probably worth it.

I'd say go for it, but maybe prototype with a standard reference model first to get your pipeline working, then swap in RLT once you've proven the concept.

**Comment by kiplangatkorir** [comment]
> Thank you so so much <3!

You are welcome brother.

**Comment by kiplangatkorir** [comment]
Oh man, this is fascinating! This completely flips my earlier advice on its head.

So basically what they're showing is that RLVR might just be a fancy way of "waking up" knowledge that's already baked into the model from pretraining, rather than actually teaching it new stuff. That's a game-changer for your phrase generation project.

If spurious rewards can get you 21-24% improvement vs 29% with perfect ground truth, then yeah, maybe you don't need to spend months curating perfect phrase labels. You could probably get away with much simpler signals like:
- Does it follow the format?
- Is the phrase length reasonable?
- Maybe some basic keyword overlap checks

The model might already "know" what good research phrases look like from seeing tons of papers during pretraining. RLVR just helps surface that knowledge.

But the model dependency thing is crucial - this worked great on Qwen but failed on other families. So you'd need to pick your base model carefully. If you're going with Qwen, this could save you tons of annotation work.

The "code reasoning" activation pattern they found is interesting too. Makes me wonder what other latent capabilities get unlocked during RLVR that weren't obviously related to the reward signal.

This definitely makes your RLT idea even more appealing - you could combine a lightweight reward signal with the RLT approach and potentially get really good results without the massive data requirements I was pushing earlier.

What base model were you thinking of using?

**Comment by Skimmable-Code-pls** [comment]
@kiplangatkorir That depends on my CEO's choice. I will present 2 course of actions:

1. The long way: Use Llama-3.1-8B-Instruct - will need to pretrain on Math long-CoT; then pretrain on Coding long-CoT just to get to Qwen-7B-Math level; then pretrain on `short_phrase` long-CoT
2. The short way: Use Qwen-7B-Math - will only need to pretrain on 1k-samples Math long-CoT; then pretrain on 1k-samples Coding long-CoT; then pretrain on `short_phrase` long-CoT

**Comment by kiplangatkorir** [comment]
Smart framing for the CEO! The cost-benefit is pretty clear when you lay it out like that.

For the long way with Llama, yeah you're probably looking at something like the Mixture-of-Thoughts scale - so 100k+ samples for each domain to get decent coverage. That's a massive data collection and compute investment just to get to baseline Qwen performance.

The short way makes way more sense, especially given what we just learned from that spurious rewards paper. Qwen already has the math reasoning baked in, so you're just doing targeted fine-tuning rather than building from scratch.

For the `short_phrase` long-CoT part, I'd actually start much smaller than you might think. Maybe 500-1k samples to begin with, then see how it performs. The spurious rewards finding suggests the model might already understand research terminology patterns from pretraining.

Your CoT for phrase generation could be something like:
```
"Let me analyze this paper... The title mentions X, the abstract focuses on Y, existing phrases include Z... This seems most similar to [existing phrase] because... Therefore I'll use EXISTING: [phrase]"
```

But honestly, given the efficiency gains and the fact that you're doing a pretty specialized task, I'd lean heavily toward recommending option 2 to your CEO. The ROI is just way better - you get to market faster, spend way less on compute, and can iterate more quickly.

Plus if it doesn't work out, you haven't burned through your budget on the long approach.
